enhancements/baremetal/baremetal-ironic-config-passthrough.md

```diff
@@ -0,0 +1,140 @@
+---
+title: baremetal-ironic-config-passthrough
+authors:
+  - "@derekh"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-07-17
+last-updated: 2020-07-17
+status: provisional
+see-also:
+  - "https://github.com/openshift/installer/pull/3887"
+  - "https://github.com/openshift/machine-api-operator/pull/645"
+---
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Baremetal IPI makes use of OpenStack Ironic to provision RHCOS to each baremetal
+host. Ironic is deployed in the metal3 pod and its configuration is highly opinionated.
+This enhancement would provide a mechanism to pass arbitary ironic config options
+into the metal3 pod in order to change how ironic and ironic-inspector are configured.
+
+## Motivation
+
+The current opinionated ironic works well when we have a single set of target hardware
+to support, this is the ideal but hasn't been our experience. As users deploy IPI clusters
+in labs, production systems, and for CI, we occasionally needs to deal with corner cases arise.
+
+The motivation for this enhancement is to allow people to deal with their individual corner
+cases without the need to insert special cases into the ironic containers. The ironic
+container can stay opinionated and targeted for a specific set of environments while also
+being usable where uniq envirnments require adjustments.
```

Comment:
Do you have some examples of configuration changes that users have needed to change?

Reply:
Yup, there are 3 examples ongoing that I can think of 

Operators trying to deploy kni on openstack(for demos) have been looking for a way to customise IPMI timouts and retries, see here
https://github.com/metal3-io/ironic-image/pull/167

For CI, we've been trying to find a way to add certain kernel args for CI only, one of this merged which I believe means IPA is no longer being logged to the console, making debugging very difficult
https://github.com/metal3-io/ironic-image/pull/172
https://github.com/metal3-io/ironic-image/pull/160

For debugging IPA users need to add a SSH key to kernel args (this one could probbaly be a top level config option)
https://github.com/metal3-io/ironic-image/pull/150

Reply:
> Yup, there are 3 examples ongoing that I can think of
> 
> Operators trying to deploy kni on openstack(for demos) have been looking for a way to customise IPMI timouts and retries, see here
> [metal3-io/ironic-image#167](https://github.com/metal3-io/ironic-image/pull/167)

That seems like a reasonable thing to put into an API. Is it configurable on a per-host basis, or do we have to put a value into the ironic config file?

> For CI, we've been trying to find a way to add certain kernel args for CI only, one of this merged which I believe means IPA is no longer being logged to the console, making debugging very difficult
> [metal3-io/ironic-image#172](https://github.com/metal3-io/ironic-image/pull/172)
> [metal3-io/ironic-image#160](https://github.com/metal3-io/ironic-image/pull/160)

Is there any reason we wouldn't want serial console logging all the time? Can we re-enable it and write a test to ensure it can't be disabled accidentally?

If it needs to be configurable, is it a per-host option or global?

> For debugging IPA users need to add a SSH key to kernel args (this one could probbaly be a top level config option)
> [metal3-io/ironic-image#150](https://github.com/metal3-io/ironic-image/pull/150)

Could we use the same ssh key that is given in the install-config for access to the control plane nodes?